### PR TITLE
Implement retry mechanism with exponential backoff for feed fetching

### LINKED ISSUE #19: Add retry mechanism with exponential backoff
## Description
Implement retry logic with exponential backoff for transient network failures.

## Current Implementation
- No retry mechanism in `store/store.go:64`
- Single failure marks feed as permanently failed until cache expires

## Proposed Changes
1. Add retry package (e.g., `github.com/cenkalti/backoff`)
2. Configure retry attempts and backoff intervals
3. Distinguish between retryable and non-retryable errors
4. Add jitter to prevent thundering herd

## Success Criteria
- [ ] Configurable retry attempts (default: 3)
- [ ] Exponential backoff with jitter
- [ ] HTTP 5xx errors trigger retries, 4xx errors don't
- [ ] Timeout errors trigger retries
- [ ] DNS errors trigger retries
- [ ] Tests verify retry behavior and backoff timing
- [ ] Metrics track retry attempts and success rates

## Implementation Notes
- Distinguish permanent failures (404, invalid feed) from temporary (503, timeout)
- Consider max total retry time limit

### PR DESCRIPTION
**Pull request by richardwooding**
## Summary

This PR implements a comprehensive retry mechanism with exponential backoff for feed fetching, addressing issue #19. The implementation provides intelligent error classification, configurable retry parameters, and comprehensive metrics tracking.

### Core Implementation
- **Configurable retry attempts** (default: 3, CLI: `--retry-max-attempts`)
- **Exponential backoff with jitter** (base: 1s, max: 30s, configurable)
- **Intelligent error classification** (5xx/timeout/DNS errors retry, 4xx don't)
- **Context-aware cancellation** support for graceful shutdown
- **Thread-safe metrics collection** with sync.RWMutex

### CLI Configuration
```bash
--retry-max-attempts=3        # Maximum retry attempts
--retry-base-delay=1s         # Base delay for exponential backoff  
--retry-max-delay=30s         # Maximum delay cap
--retry-jitter=true           # Enable jitter to prevent thundering herd
```

### Error Classification Logic
- **HTTP 5xx server errors**: Retryable (temporary server issues)
- **HTTP 4xx client errors**: Non-retryable (permanent client issues) 
- **DNS/network errors**: Retryable (temporary connectivity issues)
- **Context cancellation**: Non-retryable (graceful shutdown)

### Metrics Tracking
- Total attempts and retries across all feeds
- Success/failure rates for monitoring and tuning
- Per-feed retry statistics via `GetRetryMetrics()`

### Integration
- Seamlessly works with existing circuit breaker and connection pooling
- Preserves all existing functionality and API compatibility
- No breaking changes to public APIs

### Files Modified
- **store/store.go**: Core retry logic, error classification, and metrics
- **cmd/cmd.go**: CLI configuration flags for retry parameters
- **store/store_test.go**: Comprehensive test suite (15 new tests)
- **README.md & CLAUDE.md**: Updated documentation with retry details
- **cmd/*_test.go**: Test fixes for proper component isolation

### Test Coverage
- **7 retry mechanism tests**: Various failure scenarios and timing verification
- **4 metrics tests**: Success/failure tracking across different scenarios  
- **Timing tests**: Exponential backoff delay verification
- **Error classification tests**: All error types properly classified
- **Integration tests**: Works correctly with circuit breakers and connection pooling

### All Tests Passing
```bash
$ go test ./...
ok      github.com/richardwooding/feed-mcp      (cached)
ok      github.com/richardwooding/feed-mcp/cmd  0.203s
ok      github.com/richardwooding/feed-mcp/mcpserver    (cached)
ok      github.com/richardwooding/feed-mcp/model        (cached)
ok      github.com/richardwooding/feed-mcp/store        (cached)
```

### Basic Usage with Defaults
```bash
# Uses default retry settings (3 attempts, 1s base delay, 30s max)
go run main.go run https://techcrunch.com/feed/
```

### Custom Retry Configuration
```bash
# Custom retry configuration
go run main.go run \
  --retry-max-attempts=5 \
  --retry-base-delay=2s \
  --retry-max-delay=60s \
  --retry-jitter=false \
  https://example.com/feed/
```

## 🎯 Resolves

Closes #19

## Test plan

- [x] All existing tests pass
- [x] New retry mechanism tests pass (7 tests)
- [x] New metrics tests pass (4 tests) 
- [x] Manual testing with failing/slow feeds
- [x] Integration testing with circuit breakers
- [x] CLI flag validation and defaults
- [x] Documentation updated and verified

🤖 Generated with [Claude Code](https://claude.ai/code)

### CODE REVIEW

File: store/store_test.go [hunk 1]
```diff
@@ -524,13 +524,14 @@ func TestStore_CircuitBreakerRecovery(t *testing.T) {
 	}))
 	defer recoveringServer.Close()
 
-	// Create store with circuit breaker enabled
+	// Create store with circuit breaker enabled but retries disabled for predictable testing
```
Comment (richardwooding) [inline store/store_test.go:527]: ✅ Fixed! Thank you for catching the grammatical error. Updated to "retries disabled" for correct plural form.

File: store/store.go [hunk 2]
```diff
@@ -153,6 +320,18 @@ func NewStore(config Config) (*Store, error) {
 		config.IdleConnTimeout = 90 * time.Second // Default to 90 seconds idle timeout
 	}
 
+	// Set default retry values
+	if config.RetryMaxAttempts <= 0 {
```
Comment (richardwooding) [inline store/store.go:324]: ✅ Good point! I clarified the comment to explicitly reference that the default is handled by the CLI flag. The Kong framework sets `RetryJitter` to `true` by default via the struct tag `default:"true"`, so no additional code logic is needed here. The comment now accurately reflects this implementation pattern.

File: store/store.go [hunk 3]
```diff
@@ -102,6 +122,153 @@ func NewRateLimitedHTTPClient(requestsPerSecond float64, burstCapacity int, pool
...
+
+	// Cap at maxDelay
+	if delay > maxDelay {
+		delay = maxDelay
+	}
+
+	// Add jitter to avoid thundering herd
+	if useJitter && delay > 0 {
+		jitterRange := delay / 2
+		var jitter time.Duration
```
Comment (richardwooding) [inline store/store.go:178]: ✅ Excellent catch! Added proper safety checks for the jitter calculation:
- Check if `jitterRange > 0` before calling `rand.Int63n()` to prevent panic
- Added bounds checking to ensure the final delay is never negative
- This handles edge cases with very small delay values safely

The jitter logic now robustly handles all edge cases while maintaining the intended behavior.

File: store/store.go [hunk 4]
```diff
@@ -102,6 +122,153 @@ func NewRateLimitedHTTPClient(requestsPerSecond float64, burstCapacity int, pool
...
+		if jitterRange > 0 {
+			jitter = time.Duration(rand.Int63n(int64(jitterRange)))
```
Comment (richardwooding) [inline store/store.go:180]: ✅ Implemented both suggestions! The code now includes:

1. Safe jitter calculation with proper bounds checking for `jitterRange > 0`
2. Final delay validation to ensure it's never negative

This prevents both the `rand.Int63n(0)` panic and negative delay edge cases. The implementation maintains the intended exponential backoff behavior while being robust against all edge cases.

**Review by richardwooding — COMMENTED** [review]
Thank you @Copilot for the thorough code review! 🤖

All feedback has been addressed in commit adb8c24:

## Summary of Changes
✅ **Grammar fix**: Updated test comment to use correct plural "retries disabled"
✅ **Comment clarity**: Clarified that RetryJitter defaults are handled by CLI flag defaults  
✅ **Safety improvements**: Added comprehensive bounds checking for jitter calculations

## Technical Details
The jitter calculation now includes robust error handling:
- Prevents `rand.Int63n(0)` panics with zero-value checks
- Ensures final delay values are never negative
- Maintains intended exponential backoff behavior across all edge cases

These changes improve code reliability without affecting functionality. All existing tests continue to pass, confirming backward compatibility.
